src/lib/common/sol-platform-impl-linux-micro.c

```diff
@@ -514,16 +522,137 @@ teardown_pid1(void)
+        return false;
+    }
+
+    file = sol_file_reader_get_all(fr);
+    // remove the \n in the end of the file (this is a single line file - always)
+    file.len--;
+    tokens = sol_util_str_split(file, " ", 0);
+    SOL_VECTOR_FOREACH_IDX (&tokens, token, i) {
+        char arg[PATH_MAX] = { 0 };
+        int err;
+
+        err = snprintf(arg, sizeof(arg), "%.*s", SOL_STR_SLICE_PRINT(*token));
+        if (err < 0 || err >= (int)sizeof(arg)) {
+            SOL_ERR("Could not format arg");
+            continue;
+        }
+
+        if (streq(arg, SOL_DEBUG_ARG)) {
+            res = true;
+        } else if (strstartswith(arg, SOL_DEBUG_COMM_ARG)) {
+            gdb->comm = strdup(arg);
+            if (res) {
+                SOL_DBG("Got "SOL_DEBUG_ARG " flag and comm: %s", arg);
+                break;
+            }
+        }
+    }
+
+    // if no comm is provided set the default one
+    if (!gdb->comm) {
+        gdb->comm = strdup("/dev/tty0");
+        if (!gdb->comm) {
+            SOL_INF("Could not allocate gdbserver comm desc memory");
+            return false;
+        }
+        SOL_INF("No debug comm provided, using default one (/dev/tty0)");
+    }
+
+    sol_file_reader_close(fr);
+    return res;
+}
+
+static void
+on_fork(void *data)
+{
+    int err;
+    unsigned int i;
+    char pids[PATH_MAX];
+    struct gdbserver *gdb = data;
+
+    const char *paths[] = {
+        "/usr/bin/gdbserver",
+        "/bin/gdbserver",
+    };
+
+    err = snprintf(pids, sizeof(pids), "%d", gdb->pid);
+    if (err < 0 || err >= (int)sizeof(pids)) {
+        SOL_ERR("Could not format pid string to attach gdbserver");
+        sol_platform_linux_fork_run_exit(EXIT_FAILURE);
+    }
+
+    if (setenv("SOL_LOAD_INITIAL_SERVICES", "1", 1) == -1) {
+        SOL_ERR("Could not set SOL_LOAD_INITIAL_SERVICES");
+        sol_platform_linux_fork_run_exit(EXIT_FAILURE);
+    }
+
+    for (i = 0; i < ARRAY_SIZE(paths); i++) {
+        if (execl(paths[i], paths[i], "--attach", gdb->comm, pids, NULL) == -1) {
+            SOL_ERR("failed to exec gdbserver - %s", sol_util_strerrora(errno));
+            continue;
+        }
+    }
+
+    SOL_ERR("Failed to exec() gdbserver");
+    sol_platform_linux_fork_run_exit(EXIT_FAILURE);
+}
+
+static void
+on_fork_exit(void *data, uint64_t pid, int status)
+{
+    SOL_ERR("gdbserver exited... %" PRIu64, pid);
+}
+
 int
 sol_platform_impl_init(void)
 {
     bool want_load_initial_services = false;
+    struct gdbserver gdb;
+    pid_t pid = getpid();
 
-    if (getpid() == 1 && getppid() == 0) {
+    if (pid == 1 && getppid() == 0) {
         int err = setup_pid1();
         SOL_INT_CHECK(err, < 0, err);
 
         want_load_initial_services = true;
+
+        gdb.pid = pid;
+        gdb.comm = NULL;
+
+        if (sol_platform_linux_micro_should_debug(&gdb)) {
+            struct sol_platform_linux_fork_run *fork;
+
+            fork = sol_platform_linux_fork_run(on_fork, on_fork_exit, &gdb);
```

Comment:
this is not required, it will do some stuff you don't need, and while it seem to help, it doesn't much and it will introduce more code paths (like malloc and adding a child to be monitored by the main loop -- that is not going to run).

do the fork yourself, no need for synchronization pipes, but I did forget to reset the sigprocmask in my version. This way we reduce code paths for this case.